### PR TITLE
Add Swift Package Manager support

### LINKED ISSUE #1176: Add Swift Package Manager support
Related to #1149

> The following PRs need to be merged first before being able to add a `Package.swift` manifest to `XMPPFramework`
> - KissXML: [#119](https://github.com/robbiehanson/KissXML/pull/119)
> - libidn-framework: [#6](https://github.com/chrisballinger/libidn-framework/pull/6)

### PR DESCRIPTION
**Pull request by karimhm**
Fixes #1176

### CODE REVIEW

**Comment by karimhm** [comment]
⚠️ This PR is half done. `XMPPFramework` will crash when attempting to use Core Data storage. ⚠️

As of Swift v5.2 it is not possible to distribute resources alongside a Swift Package, therefore the Core Data models need to be copied in a way or another as they are needed for the proper function of `XMPPFramework`.

They can be copied in serval ways:
- Copy them one by one inside the target application.
- Group all the models inside a bundle and then that bundle needs to be copied into the target application.
- Programmatically create the models and avoid the resource distribution issue.

> This remains the case until Xcode 12 becomes publicly available. 
> Please take a look at [Package Manager Resources](https://developer.apple.com/documentation/swift_packages/bundling_resources_with_a_swift_package) and [SE-0271](https://github.com/apple/swift-evolution/blob/master/proposals/0271-package-manager-resources.md).

File: Package.swift
Comment (karimhm) [inline Package.swift:25]: Those need to be replaced once **KissXML: [#119](https://github.com/robbiehanson/KissXML/pull/119)** and **libidn-framework: [#6](https://github.com/chrisballinger/libidn-framework/pull/6)** get merged.

File: Package.swift
Comment (karimhm) [inline Package.swift:76]: The following tests are excluded because they rely on Core Data storage.

**Comment by chrisballinger** [comment]
@karimhm This is awesome! What do you think about attempting a parallel `Package@swift-5.3.swift` that includes the Core Data resources until Xcode 12 reaches GM? Travis has support for Xcode 12 beta now which could allow running those tests on CI.

**Comment by karimhm** [comment]
@chrisballinger Swift tools version 5.3 are now supported via a parallel `Package .swift`
But i couldn't find a way to copy Core Data resources to the unit test bundle, otherwise everything work as expected.

**Comment by karimhm** [comment]
I posted a [question](https://forums.swift.org/t/swift-package-resources-not-copied-in-unit-tests/38840) at the swift forum regarding the issue.

**Comment by karimhm** [comment]
> @chrisballinger Swift tools version 5.3 are now supported via a parallel `Package.swift`
> But I couldn't find a way to copy Core Data resources to the unit test bundle, otherwise, everything works as expected.

This seems to be a known bug ([SR-12912](https://bugs.swift.org/browse/SR-12912)).
We need to wait until it gets resolved.

**Comment by chrisballinger** [comment]
It looks like SR-12912 won't be fixed in time for Swift 5.3, so perhaps we should move forward with this and just make sure to exclude any resource-related tests from SPM. We still have tests for those in the other Xcode projects that are run on CI. We should just make another ticket to track re-enabling those tests in SPM whenever it's fixed.

I did have another concern about the approach of using symlinks for the `include/XMPPFramework` folder. It seems like there will be a maintenance burden of maintaining those symlinks when new files are added to the project, or moved around. Perhaps it would be better to reorganize the project's folder structure to avoid needing the symlinks.

**Comment by karimhm** [comment]
This PR have unit tests on the SPM side excluding files that depend on module resources. Until SR-12912 gets resolved.

**Comment by karimhm** [comment]
This might be related to [SR-13560](https://bugs.swift.org/browse/SR-13560), who knows.

**Comment by karimhm** [comment]
[SR-13560](https://bugs.swift.org/browse/SR-13560) seems to be the bug that is preventing tests from running properly.

**Comment by karimhm** [comment]
@chrisballinger All the tests are running now. Nothing excluded.

File: .travis.yml
Comment (chrisballinger) [inline .travis.yml:12]: Great job on the workaround! Hopefully Apple fixes this bug someday. Is there a [Feedback Assistant](https://feedbackassistant.apple.com) report number that could be added to that ticket?

**Review by chrisballinger — CHANGES_REQUESTED** [review]
@karimhm This is awesome stuff! Will be good to merge once the package manifest dependency URLs are fixed.

On a side note, Travis-CI has been crippled for open source projects and it would make sense to migrate to GitHub Actions at some point soon.

File: .travis.yml
Comment (karimhm) [reply .travis.yml:12]: Im not aware of any

**Comment by karimhm** [comment]
> @karimhm This is awesome stuff! Will be good to merge once the package manifest dependency URLs are fixed.
> 
> On a side note, Travis-CI has been crippled for open source projects and it would make sense to migrate to GitHub Actions at some point soon.

This can be done once this PR get merged

**Review by chrisballinger — APPROVED** [review]

**Pull request merged**
